src/gc/gc.cpp

```diff
@@ -34212,11 +34212,30 @@ HRESULT GCHeap::Initialize()
 
 #ifdef MULTIPLE_HEAPS
     nhp_from_config = static_cast<uint32_t>(GCConfig::GetHeapCount());
-    
+
     // GetCurrentProcessCpuCount only returns up to 64 procs.
-    uint32_t nhp_from_process = GCToOSInterface::CanEnableGCCPUGroups() ?
-                                GCToOSInterface::GetTotalProcessorCount():
-                                GCToOSInterface::GetCurrentProcessCpuCount();
+    uint32_t nhp_from_process;
+    if (GCToOSInterface::CanEnableGCCPUGroups())
+    {
+        nhp_from_process = GCToOSInterface::GetTotalProcessorCount();
+    }
+    else
+    {
+        uint32_t cpuCount = GCToOSInterface::GetCurrentProcessCpuCount();
+        double cpuBudget = GCToOSInterface::GetCurrentProcessCpuBudget();
```

Comment:
the cpuBudget does not belong to GC - GC shouldn't even need to care about the concept of a cpu budget. this should part of GetCurrentProcessCpuCount implementation that GC calls.

Reply:
I did that following https://github.com/dotnet/coreclr/pull/23398#issuecomment-475664977. Are you suggesting I do this computation in `GCToOSInterface::GetCurrentProcessCpuCount` instead?

Reply:
correct. 'cause GC is not doing anything different when there's this cpu budget vs it's set via other ways (ie, all it cares about is to compute a processor count from the `GCToOSInterface` API). if we were to do things differently say when the processor count is set by the cpu budget vs affinity it would make sense to have separate APIs but since we are not, I don't see any reason why the cpu budget concept needs to be known by the GC.

Reply:
@luhenry this seems to be unrelated to my comment you've referenced. I was saying that it feels strange to expose different CPU counts to different audiences (GC, threadpool, users of Environment.ProcessorCount). Now whether this is done at this place or in the GCToOSInterface::GetCurrentProcessCpuCount, it exposes the same thing to GC.
Also, my comment is from the time before @VSadov commented on the difference of meaning of CPU count to different components (threadpool wanting the real number of cores, parallel.foreach rather the "budget").

Reply:
There are multiple uses of `GCToOSInterface::GetCurrentProcessCpuCount`:
 - https://github.com/dotnet/coreclr/blob/a28b25aacdcd2adb0fdfa70bd869f53ba6565976/src/gc/handletablecache.cpp#L60
 - `gc.cpp:34219`

I have taken this approach for two reasons: 1. for consistency between all various callers of `GetCurrentProcessCpuCount`, and 2. it's still possible to have more cores than CPU limit so it might be useful - for performance purposes - to keep the distiction.

How much do you care for these two above reasons? If not that much, then I'll go ahead and put the call to `::GetCurrentProcessCpuBudget` in `GCToOSInterface::GetCurrentProcessCpuCount`.

Reply:
@Maoni0 with my upcoming GC changes related to the CPU groups, this code would go behind the GCToOSInterface anyways. I leave it up to you to decide whether the `static uint32_t uNonSleepSpins = 8 * (GCToOSInterface::GetCurrentProcessCpuCount() - 1); ` should use the real number of CPUs available for running the process or the "budget".

Reply:
I would go with the same implementation I suggested for GC for the handle table usage (uNonSleepSpins).

Reply:
I put all the code in `GCToOSInterface::GetCurrentProcessCpuCount`